### DISCUSSION TITLE
Force perfect items with debug command drop

### DISCUSSION THREAD

**Discussion by malvarenga123** [diasurgical/DevilutionX#6211] [Q&A]
Hello, is there a way to force items dropped using the "drop" command in debug builds to have their maximum attribute value?
For example, an Obsidian Ring of the Zodiac with +20 to all attributes and +40% all resists and not a random one with +18/+39%.

**Comment by qndel** [comment]
no

**Comment by DakkJaniels** [comment]
Yes, change the code to not drop an item unless it's perfect, then rebuild it.

**Comment by malvarenga123** [comment]
Thanks @StephenCWills

**Reply by Gerwin2k** [reply]
I was experimenting a bit with the "drop" debug command yesterday. 
There are a few popular items that are barely generating within the 1.000.000 attempts. Like "Awesome plate of the stars" or "Massive Bow of Swiftness". So forcing a perfect variant filter over that would make it fail most of the time, and make it appear unresponsive. 
AFAIK: Some items may not even exist as a legit 100% perfect item: For example, the "Awesome plate of the stars" variant of the Gothic Plate, which I have so far not seen as the max. 150% armor together with max. +11 all attributes.

**Reply by malvarenga123** [reply]
Weird. @StephenCWills's answer disappeared. But I built a debug version with the "dropp" and I even increased the attempts to 9.999.999 and perfect versions of some items simply do not drop. Awesome Plate of the Stars is one of them. Godly Plate of the Whale is another.
I was able to get perfect King's Sword of Haste and Obsidian Rings and Amulet of the Zodiac.

Edit: I think the PR was deleted. I can't find it by searching for "dropp"

**Reply by StephenCWills** [reply]
The author nuked his PR because we started discussing methods for detecting cheats in Discord, and he began to feel bad about having created it.

**Reply by malvarenga123** [reply]
But it was related to a debug build. Most debug commands are supposed to be "cheats".

**Reply by DakkJaniels** [reply]
You know you can put debug code in a release build, right? Regardless it's the author's choice.

**Reply by DakkJaniels** [reply]
Regarding perfect items not dropping, with the drop command it also randomizes the item index, but the seed that generated that item index is not saved. For dungeon drops, the game uses the item's index and then regenerates the affixes from the seed, so randomizing the item index will end up generating a lot of items that you aren't looking for (depending on the item requested, you are probably looking at <10% chance of even getting the item type correct). At one point I just wrote a function to allow me to just feed in the item index I was looking for, so I could get a bit better control of the items that were being generated.
